frontend/src/utilities/js-messages.ts

```diff
@@ -0,0 +1,260 @@
+
+	@TupleToVec2
+	multiplier!: { x: number; y: number };
+}
+
+export class UpdateRulers extends JsMessage {
+	@TupleToVec2
+	origin!: { x: number; y: number };
+
+	spacing!: number;
+
+	interval!: number;
+}
+
+export class ExportDocument extends JsMessage {
+	document!: string;
+
+	name!: string;
+}
+
+export class SaveDocument extends JsMessage {
+	document!: string;
+
+	name!: string;
+}
+
+export class OpenDocumentBrowse extends JsMessage {}
+
+export class DocumentChanged extends JsMessage {}
+
+export class DisplayFolderTreeStructure extends JsMessage {
+	constructor(public layerId: BigInt, public children: DisplayFolderTreeStructure[]) {
+		super();
+	}
+}
+export function newDisplayFolderTreeStructure(input: any): DisplayFolderTreeStructure {
+	const { ptr, len } = input.data_buffer;
+	const wasmMemoryBuffer = (window as any).wasmMemory().buffer;
+
+	// Decode the folder structure encoding
+	const encoding = new DataView(wasmMemoryBuffer, ptr, len);
+
+	// The structure section indicates how to read through the upcoming layer list and assign depths to each layer
+	const structureSectionLength = Number(encoding.getBigUint64(0, true));
+	const structureSectionMsbSigned = new DataView(wasmMemoryBuffer, ptr + 8, structureSectionLength * 8);
+
+	// The layer IDs section lists each layer ID sequentially in the tree, as it will show up in the panel
+	const layerIdsSection = new DataView(wasmMemoryBuffer, ptr + 8 + structureSectionLength * 8);
+
+	let layersEncountered = 0;
+	let currentFolder = new DisplayFolderTreeStructure(BigInt(-1), []);
+	const currentFolderStack = [currentFolder];
+
+	for (let i = 0; i < structureSectionLength; i += 1) {
+		const msbSigned = structureSectionMsbSigned.getBigUint64(i * 8, true);
+		const msbMask = BigInt(1) << BigInt(63);
+
+		// Set the MSB to 0 to clear the sign and then read the number as usual
+		const numberOfLayersAtThisDepth = msbSigned & ~msbMask;
+
+		// Store child folders in the current folder (until we are interrupted by an indent)
+		for (let j = 0; j < numberOfLayersAtThisDepth; j += 1) {
+			const layerId = layerIdsSection.getBigUint64(layersEncountered * 8, true);
+			layersEncountered += 1;
+
+			const childLayer = new DisplayFolderTreeStructure(layerId, []);
+			currentFolder.children.push(childLayer);
+		}
+
+		// Check the sign of the MSB, where a 1 is a negative (outward) indent
+		const subsequentDirectionOfDepthChange = (msbSigned & msbMask) === BigInt(0);
+		// debugger;
+		// Inward
+		if (subsequentDirectionOfDepthChange) {
+			currentFolderStack.push(currentFolder);
+			currentFolder = currentFolder.children[currentFolder.children.length - 1];
+		}
+		// Outward
+		else {
+			const popped = currentFolderStack.pop();
+			if (!popped) throw Error("Too many negative indents in the folder structure");
+			if (popped) currentFolder = popped;
+		}
+	}
+
+	return currentFolder;
+}
+
+// NOTE: Why was it newPath(input.data.path) instead of input.path. Seems to work at input.path?
```

Comment:
Do you know why it is this way?

Reply:
It looks like `editor\src\frontend\frontend_message_handler.rs:18` defines this as `UpdateLayer { path: Path, data: LayerPanelEntry },` and that `LayerPanelEntry` also has a `path` field. I am not sure precisely what the difference is without further digging, but I'd suggest leaving it alone to prevent a regression, unless you believe that was already a mistake. We can as TrueDoctor for his thoughts, if so.

Reply:
Currently they are the same thing, but I am actually not sure if that will always be the case.
It is basically a key value pair in which the value also contains the key. But that does not currently make a difference.

Reply:
Is one of the two a safer option to use? @TrueDoctor

Reply:
Too late rn will try not to forget to answer tororow

Reply:
I don't think there is a good reason to keep both of them in there, we should just get rid of one of them

Reply:
Ok, I removed path from the UpdateLayer Message